engine/graphics/src/opengl/graphics_opengl.cpp

```diff
@@ -2036,46 +2336,46 @@ static uintptr_t GetExtProcAddress(const char* name, const char* extension_name,
 
         OpenGLProgram* program_ptr = (OpenGLProgram*) program;
 
-        glLinkProgram(program_ptr->m_Program);
+        glLinkProgram(program_ptr->m_Id);
         CHECK_GL_ERROR;
 
         BuildAttributes(program_ptr);
 
+        BuildUniforms(program_ptr);
+
         return true;
     }
 
     static uint32_t OpenGLGetUniformCount(HProgram prog)
     {
-        GLint count;
         OpenGLProgram* program_ptr = (OpenGLProgram*) prog;
-        glGetProgramiv(program_ptr->m_Program, GL_ACTIVE_UNIFORMS, &count);
-        CHECK_GL_ERROR;
-        return count;
+        return program_ptr->m_ResourceBindings.Size();
     }
 
     static uint32_t OpenGLGetUniformName(HProgram prog, uint32_t index, char* buffer, uint32_t buffer_size, Type* type, int32_t* size)
     {
-        GLint uniform_size;
-        GLenum uniform_type;
-        GLsizei uniform_name_length;
         OpenGLProgram* program_ptr = (OpenGLProgram*) prog;
-        glGetActiveUniform(program_ptr->m_Program, index, buffer_size, &uniform_name_length, &uniform_size, &uniform_type, buffer);
-        *type = GetGraphicsType(uniform_type);
-        *size = uniform_size;
-        CHECK_GL_ERROR;
-        return (uint32_t)uniform_name_length;
+        ShaderResourceBinding& binding = program_ptr->m_ResourceBindings[index];
+
+        *type = binding.m_Type;
+        *size = binding.m_ElementCount;
+
+        return dmStrlCpy(buffer, binding.m_Name, buffer_size);
     }
 
     static int32_t OpenGLGetUniformLocation(HProgram prog, const char* name)
```

Comment:
This is called from material.cpp to map a string name to a location, but we should really not store any strings for that purpose. We can prepare all the data the material.cpp needs and pass in a hash here instead since we already know that.. It's the same for the other adapters as well, they don't use strings afaik